### PR TITLE
Pre-generate SNI test certs

### PR DESCRIPTION
**Pull request by liggitt**
Pre-generates test certs for SNI tests, since doing this dynamically can take a loooong time in entropy-starved or CPU-bound test envs (like in a container).

### CODE REVIEW

**Comment by liggitt** [comment]
cc @ncdc @sttts

File: staging/src/k8s.io/apiserver/pkg/server/options/serving_test.go [hunk 1]
```diff
@@ -663,3 +675,55 @@ func fakeVersion() version.Info {
...
+		return nil, nil, err
+	}
+
+	template := x509.Certificate{
+		SerialNumber: big.NewInt(1),
+		Subject: pkix.Name{
+			CommonName: fmt.Sprintf("%s@%d", host, time.Now().Unix()),
+		},
+		NotBefore: time.Unix(0, 0),
+		NotAfter:  time.Now().Add(time.Hour * 24 * 365 * 100),
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/server/options/serving_test.go]: inlined to generate certs valid for 100 years

**Comment by wojtek-t** [comment]
I don't have enough context to review this.

**Comment by sttts** [comment]
Can you put a Makefile next to the generated certs?

**Comment by sttts** [comment]
Oh, missed `Using pre-generated test certs, run with GENERATE_TEST_CERTS=true to regenerate` :+1:

File: staging/src/k8s.io/apiserver/pkg/server/options/serving_test.go [hunk 2]
```diff
@@ -247,6 +254,13 @@ NextTest:
 }
 
 func TestServerRunWithSNI(t *testing.T) {
+	generate := os.Getenv("GENERATE_TEST_CERTS") == "true"
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/options/serving_test.go]: I would also be fine with a `createIfNotExists` mechanism. One magic env var less.

File: staging/src/k8s.io/apiserver/pkg/server/options/serving_test.go [hunk 3]
```diff
@@ -387,19 +401,22 @@ func TestServerRunWithSNI(t *testing.T) {
 		},
 	}
 
-	tempDir, err := ioutil.TempDir("", "")
-	if err != nil {
-		t.Fatal(err)
+	specToName := func(spec TestCertSpec) string {
+		return spec.host + "+" + strings.Join(spec.names, ",") + "+" + strings.Join(spec.ips, ",")
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/options/serving_test.go]: + in file names looks strange. Underscore or - is not good enough?

Comment (ncdc) [inline staging/src/k8s.io/apiserver/pkg/server/options/serving_test.go]: Would it also be worth including `host=`, `names=`, and `ips=` in the file name?

File: staging/src/k8s.io/apiserver/pkg/server/options/serving_test.go [hunk 4]
```diff
@@ -663,3 +675,55 @@ func fakeVersion() version.Info {
 		GitTreeState: "Dirty",
 	}
 }
+
+// generateSelfSignedCertKey creates a self-signed certificate and key for the given host.
+// Host may be an IP or a DNS name
+// You may also specify additional subject alt names (either ip or dns names) for the certificate
+func generateSelfSignedCertKey(host string, alternateIPs []net.IP, alternateDNS []string) ([]byte, []byte, error) {
```
Comment (ncdc) [inline staging/src/k8s.io/apiserver/pkg/server/options/serving_test.go]: Or modify `utilcert.GenerateSelfSignedCertKey` to take in a `ttl`? It's only used in 2 places in the entire code base.

**Review by ncdc — COMMENTED** [review]
I'll backport this to my openshift pr and see how it does

File: staging/src/k8s.io/apiserver/pkg/server/options/serving_test.go (hunk 3 above)
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/server/options/serving_test.go]: changed to underscore

File: staging/src/k8s.io/apiserver/pkg/server/options/serving_test.go (hunk 2 above)
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/server/options/serving_test.go]: done

File: staging/src/k8s.io/apiserver/pkg/server/options/serving_test.go [hunk 5]
```diff
@@ -581,35 +592,31 @@ func createTestTLSCerts(spec TestCertSpec) (tlsCert tls.Certificate, err error)
...
-func createTestCertFiles(dir string, spec TestCertSpec) (certFilePath, keyFilePath string, err error) {
-	certPem, keyPem, err := utilcert.GenerateSelfSignedCertKey(spec.host, parseIPList(spec.ips), spec.names)
-	if err != nil {
-		return "", "", err
-	}
-
-	certFile, err := ioutil.TempFile(dir, "cert")
-	if err != nil {
-		return "", "", err
+func getOrCreateTestCertFiles(certFileName, keyFileName string, spec TestCertSpec) (err error) {
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/options/serving_test.go]: nit: now this is more like maybeCreateTestCertFiles. For "get" I would expect that they are returned.

**Comment by dchen1107** [comment]
Submit queue redeployed. Now only allowing PRs with 1.7 milestone. Applying the milestone because this pr was lgtmed and approved before the deadline.

**Comment by redbaron** [comment]
This breaks git clone on windows machines due to '*' in filename, created https://github.com/kubernetes/kubernetes/issues/46958

**Pull request merged**